File: dev/preview/previewctl/pkg/ssh/mock.go
// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
// Licensed under the GNU Affero General Public License (AGPL).
// See License-AGPL.txt in the project root for license information.

package ssh

import (
	"context"
	"io"

	"github.com/cockroachdb/errors"
)

var _ Client = &MockClient{}

type MockCmd struct {
	CMD    string
	STDOUT []byte
	STDERR []byte
	Err    error
}

type MockClient struct {
	Command MockCmd
}

func (m MockClient) Close() error {
	return nil
}

func (m MockClient) Run(ctx context.Context, cmd string, stdout io.Writer, stderr io.Writer) error {
	if m.Command.CMD != cmd {
		return errors.New("command not found")
	}

	_, _ = stdout.Write(m.Command.STDOUT)
	_, _ = stderr.Write(m.Command.STDERR)
	return m.Command.Err
}

File: components/server/src/test/mocks/mock.ts
/**
 * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
 * Licensed under the GNU Affero General Public License (AGPL).
 * See License.AGPL.txt in the project root for license information.
 */

export interface Mock<T extends Object> {
    override(override: Partial<T>): void;
}

export function createMock<T extends Object>(target?: T): T & Mock<T> {
    const overriddenMethods: Map<string, any> = new Map();

    function override<T>(override: Partial<T>): void {
        for (const [key, value] of Object.entries(override)) {
            overriddenMethods.set(key, value);
        }
    }

    return new Proxy(target || {}, {
        get: (target, prop) => {
            const name = String(prop);
            if (name === "override") {
                return override;
            }
            const fun = overriddenMethods.get(name);
            if (fun) {
                return fun;
            }
            return (target as any)[name];
        },
    }) as unknown as T & Mock<T>;
}
